### DISCUSSION TITLE
Error code 503 during API Request

### DISCUSSION THREAD

**Discussion by Fluxus00** [inventree/InvenTree#4482] [Q&A]
Hello Guys, I am using InvenTree and its API frequently. When I am performing an API request it happens irregularity that i get an error "503 - Service not Available" but when I check the server it still runs normal. The whole error message also contains parts of an http code containing the "Service under maintenance" website of InvenTree.    
The error seems to appear randomly, smaller scripts or single API requests working as expected for most of the time. But for my larger scripts, it is almost certain, that they get interrupted by this error. The error always appears at different parts of the script and I was not able to identify something consistent. 
Since this error is very annoying I would like to ask if this error is known to someone, how i might be able to fix it and what information i could provide to make finding a solution easier.

**Comment by eeintech** [comment]
@Fluxus00 Don't have an idea about what's happening, how about implementing re-tries in your script as a workaround, does it help?

**Comment by wolflu05** [comment]
Have you tried slowing down consecutive requests by adding some delay inbetween?

**Comment by Fluxus00** [comment]
Thank you both for your answers. I decided to use a combination of both of your suggestions. Like this:
try:
   code
except:
   time.sleep(0.05)
   same code again
I know it is not the best way to handle it but in my case it works. Furthermore i will post the whole error message here later if i encounter it again so you get a better grip of what i mean.

**Reply by eeintech** [reply]
Nice, let us know how it goes!

**Reply by Fluxus00** [reply]
@eeintech So it seems like it works, a downside is that it has to be implemented on every single request in the script, which is in my case a lot. Also i was able to reproduce the error I put in a txt file because for me it looks like the error message contains the whole http code of the "maintenance" side. 
[errormessage.txt](https://github.com/inventree/InvenTree/files/10969550/errormessage.txt)

I tried to copy and paste the whole code here but the formatting just would not fit, so I thought it this would be the next best solution.

**Reply by eeintech** [reply]
That is indeed saying the server is busy doing something, I have never had this problem developing, testing and using Ki-nTree locally or with our company instances. In any case it is better to implement delay-retry as part of your core request, which you easily re-use everywhere. You may also want to look at your server/instance resource consumption to monitor if it heats up when running your script.

**Reply by wolflu05** [reply]
May you can build a wrapper for this, or we implement retry directly in the python sdk? @matmair ?

**Reply by matmair** [reply]
Retry could lead to double entry. My suggestion is to look into why the server is not performant enough to handle the requests or slowing them down a bit. If you do a stock-remove call for example we have no way to tell if it was already done.

**Reply by wolflu05** [reply]
That's a good aspect. Maybe testing out with only delays to see if your script is doing it's requests too fast in a row.

**Reply by matmair** [reply]
In my experience with large-scale Django applications, 503 tends to turn up when the database connection is congested (Most of the time by locking write calls). Maybe look into the metrics of your database and see if something is hitting a consistent ceiling. Postgres for example can be sped up a lot by optimising its settings - I have seen databases that handled 100k rows per second in/out without any problems.

**Reply by Fluxus00** [reply]
So after testing with different scripts, amount and type of requests I can say the error did not interrupt my scripts anymore. I will close the discussion later
